linkerd/tls/src/imp/openssl.rs

```diff
@@ -0,0 +1,295 @@
+use std::{
+    pin::Pin,
+    sync::Arc,
+    task::{Context, Poll},
+};
+
+use std::str::FromStr;
+use tracing::debug;
+
+#[cfg(feature = "boring-tls")]
+use {
+    boring::{
+        fips, ssl,
+        ssl::{SslAcceptor, SslConnector, SslConnectorBuilder, SslMethod, SslVerifyMode},
+        version,
+        x509::store::X509StoreBuilder,
+    },
+    tokio_boring::SslStream,
+};
+#[cfg(not(feature = "boring-tls"))]
+use {
+    openssl::{
+        fips, ssl,
+        ssl::{Ssl, SslAcceptor, SslConnector, SslConnectorBuilder, SslMethod, SslVerifyMode},
+        version,
+        x509::store::X509StoreBuilder,
+    },
+    tokio_openssl::SslStream,
+};
+
+#[derive(Clone)]
+pub struct TlsConnector(ssl::SslConnector);
+
+impl TlsConnector {
+    #[cfg(not(feature = "boring-tls"))]
+    pub async fn connect<IO>(&self, domain: Name, stream: IO) -> Result<client::TlsStream<IO>>
+    where
+        IO: AsyncRead + AsyncWrite + Unpin,
+    {
+        let ssl = self
+            .0
+            .configure()
+            .unwrap()
+            .into_ssl(domain.as_ref())
+            .unwrap();
+        let mut s = TlsStream::new(ssl, stream);
+        match Pin::new(&mut s.0).connect().await {
+            Ok(_) => Ok(s),
+            Err(err) => Err(io::Error::new(ErrorKind::Other, err)),
+        }
+    }
+    #[cfg(feature = "boring-tls")]
+    pub async fn connect<IO>(&self, domain: Name, stream: IO) -> Result<client::TlsStream<IO>>
+    where
+        IO: AsyncRead + AsyncWrite + Unpin,
+    {
+        let conf = self.0.configure().unwrap();
+        match tokio_boring::connect(conf, domain.as_ref(), stream).await {
+            Ok(ss) => Ok(ss.into()),
+            Err(err) => {
+                let _ = err.ssl();
+                Err(io::Error::new(ErrorKind::Other, "Ble"))
+            }
+        }
+    }
+}
+
+impl From<SslConnector> for TlsConnector {
+    fn from(connector: SslConnector) -> Self {
+        Self(connector)
+    }
+}
+
+impl From<SslConnectorBuilder> for TlsConnector {
+    fn from(builder: SslConnectorBuilder) -> Self {
+        builder.build().into()
+    }
+}
+
+impl From<Arc<ClientConfig>> for TlsConnector {
+    fn from(conf: Arc<ClientConfig>) -> Self {
+        println!("Setting up ssl connector");
+        let mut builder = SslConnector::builder(SslMethod::tls()).unwrap();
+
+        match conf.0.key.clone() {
+            None => debug!("No private key provided"),
+            Some(key) => {
+                debug!("Setting private key {:?}", key);
+                builder.set_private_key(key.as_ref().0.as_ref()).unwrap();
+            }
+        }
+
+        match conf.0.cert.clone() {
+            None => debug!("No certificate provided"),
+            Some(cert) => {
+                println!("Setting certificate {:?}", cert);
+                builder.set_certificate(cert.cert.as_ref()).unwrap();
+
+                for cert in &cert.chain {
+                    println!("Adding extra chain certificate {:?}", cert);
```

Comment:
we should probably remove the `println!`s here as well (or replace them with logging)